File: langs/java/src/test/java/com/github/tetrisanalyzer/piecemove/VisitedPieceMovesTest.java
package com.github.tetrisanalyzer.piecemove;

import com.github.tetrisanalyzer.board.Board;
import com.github.tetrisanalyzer.move.Move;
import com.github.tetrisanalyzer.piece.Piece;
import com.github.tetrisanalyzer.settings.TetrisAnalyzerGameSettings;
import org.junit.Before;
import org.junit.Test;

import static com.github.tetrisanalyzer.board.Board.createBoard;
import static com.github.tetrisanalyzer.piece.Piece.createPieceS;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class VisitedPieceMovesTest {
    private Board board;
    private Piece piece;

    @Before
    public void setUp() {
        board = createBoard(6, 6);
        piece = createPieceS(new TetrisAnalyzerGameSettings());
    }

    @Test
    public void visitLeft() {
        Move move = new Move(0,0,0);
        VisitedPieceMoves visitedMoves = new VisitedPieceMoves(board, piece);
        visitedMoves.visit(getMovement(move, Direction.LEFT));

        assertFalse(visitedMoves.isUnvisited(getMovement(move, Direction.LEFT)));
        assertTrue(visitedMoves.isUnvisited(getMovement(move, Direction.RIGHT)));
        assertTrue(visitedMoves.isUnvisited(getMovement(move, Direction.DOWN)));
        assertFalse(visitedMoves.isUnvisited(getMovement(move, Direction.ROTATE)));
    }

    private Movement getMovement(Move move, Direction direction) {
        PieceMove pieceMove = new PieceMove(board, piece, move);
        return new Movement(pieceMove, direction);
    }
}

File: langs/scala/src/test/scala/nu/tengstrand/tetrisanalyzer/piecemove/VisitedPieceMovesTest.scala
package nu.tengstrand.tetrisanalyzer.piecemove

import org.junit.{Before, Test}
import nu.tengstrand.tetrisanalyzer.BaseTest
import nu.tengstrand.tetrisanalyzer.move.Move
import nu.tengstrand.tetrisanalyzer.board.Board
import nu.tengstrand.tetrisanalyzer.piece.{Piece, PieceS}

class VisitedPieceMovesTest extends BaseTest {
  private var board: Board = Board(6,6)
  private var piece: Piece = PieceS()

  @Test def visitLeftPieceS(): Unit = {
    val move = new Move(0,0,0)
    val visitedMoves = new VisitedPieceMoves(board, piece)
    visitedMoves.visit(getMovement(move, new Left))

    visitedMoves.isUnvisited(getMovement(move, new Left)) should be (false)
    visitedMoves.isUnvisited(getMovement(move, new Right)) should be (true)
    visitedMoves.isUnvisited(getMovement(move, new Down)) should be (true)
    visitedMoves.isUnvisited(getMovement(move, new Rotate)) should be (false)
  }

  private def getMovement(move: Move, direction: Direction) = {
    val pieceMove = PieceMove(board, piece, move)
    new Movement(pieceMove, direction)
  }
}